pinot-core/src/main/java/com/linkedin/pinot/core/startree/v2/builder/OnHeapSingleTreeBuilder.java

```diff
@@ -0,0 +1,153 @@
+/**
+ * Copyright (C) 2014-2018 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.startree.v2.builder;
+
+import com.linkedin.pinot.core.indexsegment.immutable.ImmutableSegment;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.commons.configuration.Configuration;
+
+import static com.linkedin.pinot.core.startree.v2.StarTreeV2Constants.*;
+
+
+/**
+ * The {@code OffHeapSingleTreeBuilder} class is the single star-tree builder that uses on-heap memory.
+ */
+public class OnHeapSingleTreeBuilder extends BaseSingleTreeBuilder {
```

Comment:
Do we really need OnHeap builder? As it will likely have a huge memory penalty.

Reply:
This is much more CPU efficient, and the implementation is much more straight-forward than the off-heap one. Can be used to build star-tree for small segments or verify the result from off-heap builder.